lang/step.md

```diff
@@ -361,8 +359,7 @@ impl<M: Memory> Machine<M> {
 
         // First evaluate the return place and remember it for `Return`. (Left-to-right!)
         let ret_place = ret.try_map(|(caller_ret_place, _abi)| {
-            // FIXME: spec Ok-wrapping of return expressions does not work in closures.
-            return self.eval_place(caller_ret_place)?;
```

Comment:
Removing the `?` also does the trick, then the `specr::hidden::ret(_)` isn't even required.

Reply:
That's a bug though, it should not work without the `?` -- this shows that Ok-wrapping is not happening.

I guess we are already doing that for a lot of other closures though... so I guess fine to do it here, but something we should revisit at some point. Right now we have Ok-wrapping for functions but not closures and that is somewhat odd.

Reply:
Yeah, it's imperfect - I agree.

Note that it working without `?` is not a bug.
`specr::hidden::ret(_)` will wrap a `T` to become `NdResult<T, _>`, but will just pass through a `NdResult<T, _>` without wrapping anything.

Previously I added `specr::hidden::ret(_)` too for closures, iirc it required a few more type annotations to make minirust compile then. I think I'll add the type annotations and make a PR.

Reply:
> specr::hidden::ret(_) will wrap a T to become NdResult<T, _>, but will just pass through a NdResult<T, _> without wrapping anything.

It does? That sounds like a bug, too.^^ Also how is it doing that?

Reply:
> That sounds like a bug, too.^^

Oh, really?
It was what I intended, for what its worth ^^

Here's how it works:
https://github.com/memoryleak47/minirust-tooling/blob/main/libspecr/src/hidden/ret.rs

Alternatively we could do
```rust
trait MonadicReturn {
  type Inner;
  fn monadic_return(t: Self::Inner) -> Self;
}

impl<T, E> MonadicReturn for NdResult<T, E> {
  type Inner = T;
  fn monadic_return(t: T) -> Self { NdResult(Ok(t)) }
}
// similarly for the other monads
```
that will probably solve some of the inference problems, but then we don't get the passthrough for `NdResult<_, _>`.

Reply:
How on earth does Rust even accept those instances? Doesn't `MonadicReturn<T> for T` overlap with `MonadicReturn<T> for Option<T> `??

Anyway my intention was that in a `Result<T, E>`-returning function, I *have to* give a return expression of type `T`. That is how `try` blocks work in Rust, so this can mostly be implemented by wrapping all function bodies in a `try`, except something has to be done for early `return`s.

(And to be clear that is not a blocker for this PR since it is pre-existing, I just noticed this only now.)

Reply:
> How on earth does Rust even accept those instances? Doesn't MonadicReturn<T> for T overlap with MonadicReturn<T> for Option<T> ??

`Option<T>` just implements both `MonadicReturn<T>` and `MonadicReturn<Option<T>>`, those impls do not collide.

Reply:
> Anyway my intention was that in a Result<T, E>-returning function, I have to give a return expression of type T.

I see, I think the snippet above "Alternatively we could do" should be able to express that.
I'll open another PR for it.

Reply:
> I see, I think the snippet above "Alternatively we could do" should be able to express that.
I'll open another PR for it.

I think that will fail due to overlapping instances but I might be too pessimistic.

Reply:
Don't you still need some fallback instance for all the functions that return neither Option nor Result?

Reply:
Yep, that would only work if we are able to insert `ret` to only those functions that actually return an `NdResult` or something similar.